File: frontend/lib/store/extraction.ts
import { create } from 'zustand';
import { useAuthFetch } from './api';

interface ExtractionRequest {
  urls: string[];
  prompt?: string;
  extraction_schema?: Record<string, any>;
  enable_web_search?: boolean;
}

interface ExtractionResponse {
  job_id: string;
  status: string;
  message: string;
}

interface ExtractionStatusResponse {
  status: 'completed' | 'failed' | 'processing';
  data?: Record<string, any>;
  error?: string;
  progress?: any;
}

interface ExtractionData {
  status: 'completed' | 'failed' | 'processing' | 'pending';
  data?: Record<string, any>;
  error?: string;
  progress?: any;
  urls: string[];
}

interface ExtractionStore {
  // State
  activeExtractions: Map<string, ExtractionData>;
  isExtracting: boolean;
  error: string | null;

  // Actions
  startExtraction: (request: ExtractionRequest) => Promise<string>;
  checkExtractionStatus: (jobId: string) => Promise<ExtractionStatusResponse>;
  clearError: () => void;
  clearExtraction: (jobId: string) => void;
}

export const useExtractionStore = create<ExtractionStore>((set, get) => ({
  activeExtractions: new Map(),
  isExtracting: false,
  error: null,

  startExtraction: async (request) => {
    const authFetch = useAuthFetch();
    set({ isExtracting: true, error: null });

    try {
      const response = await authFetch('/api/v1/extraction/extract', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to start extraction');
      }

      const data: ExtractionResponse = await response.json();
      
      // Initialize extraction status
      set(state => ({
        activeExtractions: new Map(state.activeExtractions).set(data.job_id, {
          status: 'pending',
          urls: request.urls,
        }),
      }));

      return data.job_id;
    } catch (error: any) {
      set({ error: error.message, isExtracting: false });
      throw error;
    } finally {
      set({ isExtracting: false });
    }
  },

  checkExtractionStatus: async (jobId) => {
    const authFetch = useAuthFetch();
    
    try {
      const response = await authFetch(`/api/v1/extraction/extract/${jobId}`);
      
      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || 'Failed to check extraction status');
      }

      const statusData: ExtractionStatusResponse = await response.json();

      // Update extraction status in store
      set(state => {
        const currentExtraction = state.activeExtractions.get(jobId);
        if (!currentExtraction) return state;

        return {
          activeExtractions: new Map(state.activeExtractions).set(jobId, {
            ...currentExtraction,
            ...statusData,
          }),
        };
      });

      return statusData;
    } catch (error: any) {
      const errorData: ExtractionStatusResponse = {
        status: 'failed',
        error: error.message,
      };
      
      set(state => {
        const currentExtraction = state.activeExtractions.get(jobId);
        if (!currentExtraction) return state;

        return {
          activeExtractions: new Map(state.activeExtractions).set(jobId, {
            ...currentExtraction,
            ...errorData,
          }),
        };
      });

      return errorData;
    }
  },

  clearError: () => set({ error: null }),

  clearExtraction: (jobId) => set(state => {
    const newExtractions = new Map(state.activeExtractions);
    newExtractions.delete(jobId);
    return { activeExtractions: newExtractions };
  }),
})); 

File: backend/app/api/v1/endpoints/extraction.py
from fastapi import APIRouter, HTTPException, Depends, BackgroundTasks
from typing import List, Dict, Any, Optional
from pydantic import BaseModel, HttpUrl
from app.core.security import get_current_user
from app.services.extraction_service import extract_data, create_extraction_schema
from app.core.logging import setup_logger
from celery.result import AsyncResult

# Set up logging
logger = setup_logger("extraction_endpoints")

router = APIRouter()

class ExtractionRequest(BaseModel):
    """Model for extraction request"""
    urls: List[str]
    prompt: Optional[str] = None
    extraction_schema: Optional[Dict[str, Any]] = None
    enable_web_search: bool = False

class ExtractionResponse(BaseModel):
    """Model for extraction response"""
    job_id: str
    status: str
    message: str

@router.post("/extract", response_model=ExtractionResponse)
async def create_extraction(
    request: ExtractionRequest,
    current_user: Dict = Depends(get_current_user)
):
    """
    Start an extraction job
    """
    try:
        # Validate URLs
        if not request.urls:
            raise HTTPException(status_code=400, detail="No URLs provided")
            
        # Start Celery task
        task = extract_data.delay(
            urls=request.urls,
            prompt=request.prompt,
            schema=request.extraction_schema,
            enable_web_search=request.enable_web_search,
            user_id=current_user["user_id"]
        )
        
        return ExtractionResponse(
            job_id=task.id,
            status="pending",
            message="Extraction job started successfully"
        )
        
    except Exception as e:
        logger.error(f"Error creating extraction job: {str(e)}")
        logger.error(f"Request data: {request.dict()}")
        logger.error(f"Current user data: {current_user}")
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/extract/{job_id}", response_model=Dict[str, Any])
async def get_extraction_status(
    job_id: str,
    current_user: Dict = Depends(get_current_user)
):
    """
    Get the status of an extraction job
    """
    try:
        task_result = AsyncResult(job_id)
        
        if task_result.ready():
            if task_result.successful():
                return {
                    "status": "completed",
                    "data": task_result.get()
                }
            else:
                return {
                    "status": "failed",
                    "error": str(task_result.result)
                }
        else:
            return {
                "status": "processing",
                "progress": task_result.info
            }
            
    except Exception as e:
        logger.error(f"Error getting extraction status: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e)) 